### PR TITLE
List of events conn

### PR DESCRIPTION
**Pull request by mary-dcouto**

### CODE REVIEW

File: lib/ziti-tunnel-cbs/ziti_tunnel_ctrl.c [hunk 1]
```diff
@@ -802,20 +803,22 @@ static void on_enable_mfa(ziti_context ztx, int status, ziti_mfa_enrollment *enr
 
     struct ziti_instance_s *inst = ziti_app_ctx(ztx);
     mfa_event *ev = calloc(1, sizeof(struct mfa_event_s));
-    ev->operation = strdup(mfa_status_name(mfa_status_enrollment_challenge));
+    ev->operation = calloc(strlen(mfa_status_name(mfa_status_enrollment_challenge)), sizeof(char));
```
Comment (scareything) [inline lib/ziti-tunnel-cbs/ziti_tunnel_ctrl.c:806]: This is a leak. The buffer returned from calloc is never freed and the pointer to it is overwritten on the next line.

File: lib/ziti-tunnel-cbs/ziti_tunnel_ctrl.c [hunk 2]
```diff
@@ -802,20 +803,22 @@ static void on_enable_mfa(ziti_context ztx, int status, ziti_mfa_enrollment *enr
...
     }
-    ev->recovery_codes = malloc(code_len + 1);
+    ev->recovery_codes = malloc((size + 1) * sizeof(char *));
     int idx;
     for (idx=0; enrollment->recovery_codes[idx] !=0; idx++) {
         ev->recovery_codes[idx] = calloc(strlen(enrollment->recovery_codes[idx]), sizeof(char));
-        ev->recovery_codes[idx] = enrollment->recovery_codes[idx];
+        ev->recovery_codes[idx] = strdup(enrollment->recovery_codes[idx]);
     }
     ev->recovery_codes[idx] = '\0';
```
Comment (scareything) [inline lib/ziti-tunnel-cbs/ziti_tunnel_ctrl.c:823]: The nul ASCII character is not what you want here - you need to set this to NULL (or allocate the buffer with calloc).

File: lib/ziti-tunnel-cbs/ziti_tunnel_ctrl.c (hunk 1 above)
Comment (mary-dcouto) [reply lib/ziti-tunnel-cbs/ziti_tunnel_ctrl.c]: These mfa_events are free'd in the ziti-edge-tunnel.c class

File: lib/ziti-tunnel-cbs/ziti_tunnel_ctrl.c [hunk 3]
```diff
@@ -802,20 +803,22 @@ static void on_enable_mfa(ziti_context ztx, int status, ziti_mfa_enrollment *enr
 
     struct ziti_instance_s *inst = ziti_app_ctx(ztx);
     mfa_event *ev = calloc(1, sizeof(struct mfa_event_s));
-    ev->operation = strdup(mfa_status_name(mfa_status_enrollment_challenge));
+    ev->operation = calloc(strlen(mfa_status_name(mfa_status_enrollment_challenge)), sizeof(char));
+    ev->operation = mfa_status_name(mfa_status_enrollment_challenge);
     ev->operation_type = mfa_status_enrollment_challenge;
+    ev->provisioning_url = calloc(strlen(enrollment->provisioning_url), sizeof(char));
```
[the remark this answers is not in the document]
Comment (mary-dcouto) [reply lib/ziti-tunnel-cbs/ziti_tunnel_ctrl.c]: These mfa_events are free'd in the ziti-edge-tunnel.c class

File: lib/ziti-tunnel-cbs/ziti_tunnel_ctrl.c [hunk 4]
```diff
@@ -802,20 +803,22 @@ static void on_enable_mfa(ziti_context ztx, int status, ziti_mfa_enrollment *enr
...
     char **rc = enrollment->recovery_codes;
-    int code_len = 0;
+    int size = 0;
     while (*rc != NULL) {
-        code_len = code_len + strlen(*rc);
         rc++;
+        size++;
     }
-    ev->recovery_codes = malloc(code_len + 1);
+    ev->recovery_codes = malloc((size + 1) * sizeof(char *));
```
[the remark this answers is not in the document]
Comment (mary-dcouto) [reply lib/ziti-tunnel-cbs/ziti_tunnel_ctrl.c]: changed to calloc

File: programs/ziti-edge-tunnel/ziti-edge-tunnel.c [hunk 5]
```diff
@@ -406,16 +485,16 @@ static bool check_send_notification(tunnel_identity *tnl_id) {
...
     notification_message *notification = calloc(1, sizeof(struct notification_message_s));
-    notification->Message = malloc(MAXMESSAGELEN);
+    notification->Message = calloc(MAXMESSAGELEN, sizeof(char));
     if (tnl_id->MfaMaxTimeoutRem == 0) {
-        snprintf(notification->Message, MAXMESSAGELEN, "All of the services of identity %s have timed out", tnl_id->Name);
+        snprintf(notification->Message, MAXMESSAGELEN-1, "All of the services of identity %s have timed out", tnl_id->Name);
         notification->Severity = event_severity_critical;
     } else if (tnl_id->MfaMinTimeoutRem == 0) {
-        snprintf(notification->Message, MAXMESSAGELEN, "Some of the services of identity %s have timed out", tnl_id->Name);
+        snprintf(notification->Message, MAXMESSAGELEN-1, "Some of the services of identity %s have timed out", tnl_id->Name);
```
Comment (scareything) [inline programs/ziti-edge-tunnel/ziti-edge-tunnel.c:493]: subtracting 1 is not necessary. from the `snprintf` manual page:

> The snprintf() and vsnprintf() functions will write at most size-1 of the characters printed into the output string (the size'th character then gets the terminating `\0')

Comment (mary-dcouto) [reply programs/ziti-edge-tunnel/ziti-edge-tunnel.c]: removed -1 from the length field

**Pull request merged**